net: tcp2: Add support for IPv6

ozhuraki opened the pull request:
In order to support IPv4/IPv6, work with packet through
net_pkt_ip_hdr_len(), net_pkt_ip_opts_len() which account
for IPv4/IPv6, IPv4 options and IPv6 extension headers.

ozhuraki committed:
samples: net: sockets: tcp: Add a TCP sample and instructions

In order to run the TTCN-3 based sanity check, add a TCP
sample app and instructions for running the sanity check.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

pfl committed:
net: tcp2: Allocate endpoints before using

Allocate tcp2.c endpoints before storing port numbers or addresses.
Select the IPv4 source address since net_context_create_ipv4_new()
is not currently called on packet output.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

ozhuraki committed:
net: ipv4: Remove temporary interception of TCP, UDP for TCP2

Remove temporary interception of TCP, UDP for TTCN-3 based
TCP2 sanity check.

As a part of adding IPv6 support, TCP2 will register test
callbacks/inputs with net_conn_register().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Add support for a IPv6

In order to support IPv4/IPv6, work with packet through
net_pkt_ip_hdr_len(), net_pkt_ip_opts_len() which account
for IPv4/IPv6, IPv4 options and IPv6 extension headers.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Register test inputs for TTCN-3 based TCP2 sanity check

In order to support TTCN-3 based TCP2 sanity check,
register test inputs with net_conn_register() and adjust
test functions to account for this.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Use net_tcp_finalize() for checksum calculation

In order to support IPv6, use net_tcp_finalize() for checksum
calculation.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Drop obsolete ip_get(), ip6_get() macros

After adding IPv6 support, there's no need for them.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Add IPv6 support for outgoing packets

Add IPv6 support for outgoing packets.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tp: Add IPv6 support for test protocol

In order to suport using test protocol over IPv6,
add IPv6 support.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Switch from k_timer to k_delayed_work

k_timer callback is executed from ISR context, which isn't
currenty compatible with Zephyr's shell implementation.

This problem was found while testing TCP2 client side.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Don't instantiate an extra connection on connect

Don't erroneously instantiate an extra TCP connection
on TCP connect.

This problem was overlooked earlier and found while testing
TCP2 client side.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Add log_strdup() where necessary

In case the log output involves the content of the static
buffer, add log_strdup().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Guard against th_get() on NULL pkt

This problem was overlooked earlier and was found while
testing TCP2 client side.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Check the ack number in SYN_SENT

In SYN_SENT check the ack number of the incoming TCP message
against our sequence number.

This problem was overlooked earlier and was found while
testing TCP2 client side.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Don't instantiate an extra connection on SYN+ACK

On incoming SYN+ACK, use the existing TCP connection.

This problem was overlooked earlier and was found while
testing TCP2 client side.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

rlubos commented:
Is this PR ready? I wanted to test it with `echo_client`/`echo_server` with TCP2/IPv6 enabled, but both crash at startup with the following changes:
```
diff --git a/samples/net/sockets/echo_server/prj.conf b/samples/net/sockets/echo_server/prj.conf
index 81d913e90d..9bdae36376 100644
--- a/samples/net/sockets/echo_server/prj.conf
+++ b/samples/net/sockets/echo_server/prj.conf
@@ -1,9 +1,12 @@
+CONFIG_HEAP_MEM_POOL_SIZE=8192
+
 # Generic networking options
 CONFIG_NETWORKING=y
-CONFIG_NET_UDP=y
+CONFIG_NET_UDP=n
 CONFIG_NET_TCP=y
+CONFIG_NET_TCP2=y
 CONFIG_NET_IPV6=y
-CONFIG_NET_IPV4=y
+CONFIG_NET_IPV4=n
 CONFIG_NET_SOCKETS=y
 CONFIG_NET_SOCKETS_POSIX_NAMES=y
 CONFIG_POSIX_MAX_FDS=6
```

Tried with `qemu_x86`.

ozhuraki commented:
@rlubos

Thanks for the input. Do you test with E1000 against host?

The ipv6 support is ready and works against TTCN-3 sanity check and the included tcp sample.

I did a smoke test with echo_server trough SLIP against host before submitting, worked for me.

At the moment I am checking problems on the echo_client side (which we hadn’t tested before), already see problems and try to address them.

I will re-test the echo_server with E1000 with your config against host.

rlubos commented:
@ozhuraki For starters I've tested qemu vs qemu:
* `west build -b qemu_x86 -t server` on `echo_server`
* `west build -b qemu_x86 -t client` on `echo_client`

Eventually, I want to test it with wireless L2 (OpenThread/IPSP/raw 802.15.4), we should sort out the qemu first though.

ozhuraki commented:
@rlubos

> For starters I've tested qemu vs qemu

OK, thanks. I just started to check the client side and discovered that qemu vs qemu had been dropped some time ago, the documentation on this is obsolete.

The possibility is to test qemu against echo-client/echo-server from net-tools, other host tools (nc/telnet), TTCN-3.

> Eventually, I want to test it with wireless L2 (OpenThread/IPSP/raw 802.15.4), we should sort out the qemu first though.

OK. Let me check regarding the echo_server.

rlubos commented:
Just to clarify:

>  I just started to check the client side and discovered that qemu vs qemu had been dropped some time ago, the documentation on this is obsolete.

I'm quite surprised by this information. I've been using this setup for testing for quite some time now and faced no issues so far (old TCP implementation still runs). Are there plans to remove  support for this scenario, along with the following documentation section:
https://docs.zephyrproject.org/1.13.0/subsystems/networking/qemu_setup.html#network-connection-between-two-qemu-vms
?

ozhuraki committed:
samples: echo_server: Add overlays for debugging

Add overlays that might be useful for debugging:

- overlay-log.conf
   Enable debug log for socket, context, connection, TCP/TCP2.
- overlay-debug.conf
   Disable optimization and compile with debug information.
- overlay-max-stacks.conf
   Overlay to quickly sort out insufficient stack problems.
- overlay-e1000.conf

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki commented:
@rlubos

> I'm quite surprised by this information.

I was also surprised and missed that. I think we should support it, let’s check what’s missing and add it.

Regarding the echo_server. I did a smoke test with E1000 against host, works for me:

1.	echo_server configuration (I added a few overlays for troubleshooting with 5cd93a1)

```
server-debug() {
    export BUILD=${ZEPHYR_BASE}/samples/net/sockets/echo_server/build
    rm -rf $BUILD && mkdir -p $BUILD && cd $BUILD
    cmake -DBOARD=qemu_x86 \
	  -DCONFIG_NET_UDP=n \
	  -DCONFIG_NET_IPV4=n \
	  -DCONFIG_NET_TCP2=y \
	  -DCONFIG_HEAP_MEM_POOL_SIZE=8192 \
	  -DOVERLAY_CONFIG="overlay-e1000.conf;overlay-log.conf;overlay-debug.conf;overlay-max-stacks.conf" \
	  ${@} \
	  ..
    make
}
```

2.	Create TAP interface with net-tools:

```
sudo ./net-setup.sh --config zeth.conf
```

3.	Run the app
4.	Smoke test the connection:

```
telnet -6 2001:db8::2 4242
```
P.S. The echo_client is known not to work. We just started to test with it, I am investigating to get past the initial problems.

ozhuraki committed:
net: tcp2: Eliminate the need for a test receive window

In order to drop dependency on a heap, drop the test
receive window.

The receive window was needed for TTCN-3 sanity check,
but it's possible to do without it.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Eliminate the need for a test send window

In order to drop dependency on a heap, drop the
test send window.

Use net_tcp_queue_data() to receive the outgoing data
from the socket layer.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Remove obsolete test window functions

These test receive windows were used by TTCN-3 sanity check,
but it's possible to do without them.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Do not base the incoming data evaluation on PSH

Do not base the incoming data evaluation on PSH.

Also, reduce the condition branches.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Set TCP connection endpoints in net_tcp_connect()

Initialize and set IPv4/IPv6 connection endpoints in net_tcp_connect().

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
net: tcp2: Update test protocol functions

After removing the test windows, update test protocol
functions for the TTCN-3 sanity check.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

ozhuraki committed:
samples: echo_client: Add overlays for debugging

Add overlays that might be useful for debugging:

- overlay-log.conf
   Enable debug log for socket, context, connection, TCP/TCP2.
- overlay-debug.conf
   Disable optimization and compile with debug information.
- overlay-max-stacks.conf
   Overlay to quickly sort out insufficient stack problems.
- overlay-e1000.conf

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

jukkar commented on subsys/net/ip/tcp2.c:
```diff
@@ -965,6 +965,7 @@ int net_tcp_update_recv_wnd(struct net_context *context, s32_t delta)
 /* net context wants to queue data for the TCP connection */
 int net_tcp_queue_data(struct net_context *context, struct net_pkt *pkt)
 {
+	int irq_level = irq_lock();
```
What do we want to protect here, can you add more explanation to commit msg? Locking irqs seems quite heavy thing to do.

ozhuraki replied:
@jukkar 
> can you add more explanation to commit msg? Locking irqs seems quite heavy thing to do.

The message about interrupt proteciton is refined, please take a look.

pfl replied:
Will net_tcp_queue_data() actually be called from the interrupt context, or is the lock protecting simultaneous access to this function or the context?

jukkar replied:
Why mutex or semaphore is not enough here?

ozhuraki replied:
@jukkar @pfl 

You are right, a mutex is a better idea here, so droped irq locks and added mutex, please take a look.

jukkar reviewed (commented):
Lot of these commit messages were not answering the "why" question i.e., why this commit does what it does.

jukkar commented on subsys/net/ip/tcp2.c:
```diff
@@ -163,7 +163,7 @@ static char *tcp_endpoint_to_string(union tcp_endpoint *ep)
 #undef BUF_SIZE
 #undef NBUFS
-	return s;
+	return log_strdup(s);
```
This cannot be done like this. The log_strdup() call needs to be in the LOG_DBG() call, it cannot be used separately.

ozhuraki replied:
@jukkar

> The log_strdup() call needs to be in the LOG_DBG() call, it cannot be used separately.

There's no other use for these functions outside of LOG_*(), but, OK, fixed.

pfl reviewed (changes requested):
The rest looks ok, comment added.

ozhuraki committed:
net: tcp2: Add a mutex to TCP connection

The TCP connection might be concurrently modified from the
TR/TX threads, so add a mutex to protect from the concurrent
modification.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

jukkar reviewed (approved):
LGTM
